[CI] EnrichIT testSumDurationByArtist failing

**Build Scans:**
- [elasticsearch-periodic #6340 / concurrent-search-tests](https://gradle-enterprise.elastic.co/s/vfuxzi5ppf4qm)
- [elasticsearch-periodic #6277 / openjdk17_checkpart3_java-fips-matrix](https://gradle-enterprise.elastic.co/s/7lnjv6y2gmojg)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:esql:internalClusterTest" --tests "org.elasticsearch.xpack.esql.action.EnrichIT.testSumDurationByArtist" -Dtests.seed=1C2A645843890AF9 -Dtests.jvm.argline="-Des.concurrent_search=true" -Dtests.locale=xog -Dtests.timezone=HST -Druntime.java=23
```

**Applicable branches:**
8.x

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.esql.action.EnrichIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!(testSumDurationByArtist),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: 
Expected: an empty collection
     but: <[LEAK: resource was not cleaned up before it was garbage-collected.
Recent access records: 
Created at:
	in [elasticsearch[node_s1][search][T#4]][testForPushDownEnrichRule]
	org.elasticsearch.compute.operator.exchange.ExchangeSourceHandler.lambda$new$3(ExchangeSourceHandler.java:64)
	org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:217)
	org.elasticsearch.action.support.SubscribableListener$SuccessResult.complete(SubscribableListener.java:387)
	org.elasticsearch.action.support.SubscribableListener.tryComplete(SubscribableListener.java:307)
	org.elasticsearch.action.support.SubscribableListener.setResult(SubscribableListener.java:336)
	org.elasticsearch.action.support.SubscribableListener.onResponse(SubscribableListener.java:250)
	org.elasticsearch.compute.operator.exchange.ExchangeBuffer.pollPage(ExchangeBuffer.java:70)
	org.elasticsearch.compute.operator.exchange.ExchangeBuffer.discardPages(Exc
[truncated]
```

**Issue Reasons:**
- [8.x] 2 failures in test testSumDurationByArtist (0.6% fail rate in 349 executions)
- [8.x] 2 failures in pipeline elasticsearch-periodic (20.0% fail rate in 10 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/docs/changelog/122417.yaml
+++ b/docs/changelog/122417.yaml
@@ -2,4 +2,5 @@ pr: 122417
 summary: Fix listener leak in exchange service
 area: ES|QL
 type: bug
-issues: []
+issues:
+ - 122271

--- a/docs/changelog/122417.yaml
+++ b/docs/changelog/122417.yaml
@@ -1,0 +1,5 @@
+pr: 122417
+summary: Fix listener leak in exchange service
+area: ES|QL
+type: bug
+issues: []

--- a/x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/EsqlRefCountingListener.java
+++ b/x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/EsqlRefCountingListener.java
@@ -34,7 +34,8 @@ public final class EsqlRefCountingListener implements Releasable {
     }
 
     public ActionListener<Void> acquire() {
-        return refs.acquireListener().delegateResponse((l, e) -> {
+        var listener = ActionListener.assertAtLeastOnce(refs.acquireListener());
+        return listener.delegateResponse((l, e) -> {
             failureCollector.unwrapAndCollect(e);
             l.onFailure(e);
         });

--- a/x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeService.java
+++ b/x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeService.java
@@ -15,6 +15,7 @@ import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.action.ActionListenerResponseHandler;
 import org.elasticsearch.action.support.ChannelActionListener;
 import org.elasticsearch.action.support.SubscribableListener;
+import org.elasticsearch.common.breaker.CircuitBreakingException;
 import org.elasticsearch.common.component.AbstractLifecycleComponent;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
@@ -366,7 +367,13 @@ public final class ExchangeService extends AbstractLifecycleComponent {
             final long reservedBytes = allSourcesFinished ? 0 : estimatedPageSizeInBytes.get();
             if (reservedBytes > 0) {
                 // This doesn't fully protect ESQL from OOM, but reduces the likelihood.
-                blockFactory.breaker().addEstimateBytesAndMaybeBreak(reservedBytes, "fetch page");
+                try {
+                    blockFactory.breaker().addEstimateBytesAndMaybeBreak(reservedBytes, "fetch page");
+                } catch (Exception e) {
+                    assert e instanceof CircuitBreakingException : new AssertionError(e);
+                    listener.onFailure(e);
+                    return;
+                }
                 listener = ActionListener.runAfter(listener, () -> blockFactory.breaker().addWithoutBreaking(-reservedBytes));
             }
             transportService.sendChildRequest(

--- a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeListener.java
+++ b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeListener.java
@@ -47,7 +47,8 @@ final class ComputeListener implements Releasable {
      * Acquires a new listener that doesn't collect result
      */
     ActionListener<Void> acquireAvoid() {
-        return refs.acquire().delegateResponse((l, e) -> {
+        var listener = ActionListener.assertAtLeastOnce(refs.acquire());
+        return listener.delegateResponse((l, e) -> {
             try {
                 runOnFailure.run();
             } finally {

